x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtValidateUtil.java

```diff
@@ -0,0 +1,359 @@
+                    + nbf
+                    + "] > now ["
+                    + now.getTime()
+                    + "ms/"
+                    + now
+                    + "] with skew ["
+                    + (allowedClockSkewSeconds * 1000L)
+                    + "ms]."
+            );
+        }
+    }
+
+    public static void validateExpiredTime(final SignedJWT jwt, final Date now, final long allowedClockSkewSeconds) throws Exception {
+        JwtValidateUtil.validateExpiredTime(jwt.getJWTClaimsSet().getExpirationTime(), now, allowedClockSkewSeconds);
+    }
+
+    // package private, so this logic can be called from unit tests without constructing a SignedJWT
+    static void validateExpiredTime(final Date exp, final Date now, final long allowedClockSkewSeconds) throws Exception {
+        if (exp == null) {
+            throw new Exception("Invalid exp [null].");
+        } else if (now == null) {
+            throw new Exception("Invalid now [null].");
+        } else if (allowedClockSkewSeconds < 0L) {
+            throw new Exception("Invalid allowedClockSkewSeconds [" + allowedClockSkewSeconds + "] < 0.");
+        }
+        // skewSec=0 now=2:59:59.999 exp=3:00:00.000 --> pass
+        // skewSec=0 now=3:00:00.000 exp=3:00:00.000 --> fail
+        // skewSec=1 now=3:00:00.000 exp=3:00:00.000 --> pass (subtract skew from now)
+        if (now.getTime() - (allowedClockSkewSeconds * 1000L) >= exp.getTime()) {
+            throw new Exception(
+                "Invalid exp ["
+                    + exp.getTime()
+                    + "ms/"
+                    + exp
+                    + "] < now ["
+                    + now.getTime()
+                    + "ms/"
+                    + now
+                    + "] with skew ["
+                    + (allowedClockSkewSeconds * 1000L)
+                    + "ms]."
+            );
+        }
+    }
+
+    /**
+     * Look through each JWK in the JWKSet to see if they can validate the Signed JWT signature.
+     * Apply JWT kid and JWT alg filters to the JWKs to skip unnecessary signature checking.
+     *
+     * If JWT kid is present, and any JWK kid matches, only use the matching subset of JWKs. Ignore the rest.
+     * Note: JWK kid should be unique. However, this method does not assume they are unique. Each match will be tried.
+     *
+     * Depending on the JWT alg, certain HMAC/RSA/EC JWKs can be excluded.
+     * HMAC JWKs that do not meet the minimum length requirement are ignored.
+     * RSA JWKs that do not meet the minimum length requirement are ignored.
+     * EC JWKs that do not meet the exact curve requirement are ignored.
+     *
+     * @param jwt Signed JWT to be validated.
+     * @param jwkSet JWKSet of HMAC/RSA/EC JWKs. At least one JWK is required to succeed.
+     * @throws Exception Error if JWKs fail to validate the Signed JWT.
+     */
+    public static void validateSignature(final SignedJWT jwt, final JWKSet jwkSet) throws Exception {
```

Comment:
Thanks for the iteration on this method, I think this is much clearer and cleaner than the previous version. Easier to read, validate and extend.